invoker/core/src/main/java/com/google/cloud/functions/invoker/TypedFunctionExecutor.java

```diff
@@ -0,0 +1,179 @@
+  /**
+   * Returns the {@code T} of a concrete class that implements {@link BackgroundFunction
+   * BackgroundFunction<T>}. Returns an empty {@link Optional} if {@code T} can't be determined.
+   */
+  static Optional<Type> handlerTypeArgument(Class<? extends TypedFunction<?, ?>> functionClass) {
+    return Arrays.stream(functionClass.getMethods())
+        .filter(method -> method.getName().equals(APPLY_METHOD) && method.getParameterCount() == 1)
+        .map(method -> method.getGenericParameterTypes()[0])
+        .findFirst();
+  }
+
+  /** Executes the user's method, can handle all HTTP type methods. */
+  @Override
+  public void service(HttpServletRequest req, HttpServletResponse res) {
+    HttpRequestImpl reqImpl = new HttpRequestImpl(req);
+    HttpResponseImpl resImpl = new HttpResponseImpl(res);
+    ClassLoader oldContextClassLoader = Thread.currentThread().getContextClassLoader();
+
+    try {
+      Thread.currentThread().setContextClassLoader(function.getClass().getClassLoader());
+      handleRequest(reqImpl, resImpl);
+    } finally {
+      Thread.currentThread().setContextClassLoader(oldContextClassLoader);
+      resImpl.flush();
+    }
+  }
+
+  private void handleRequest(HttpRequest req, HttpResponse res) {
+    WireFormat format = configuration.getFormat();
+
+    Object reqObj = null;
+    try {
+      reqObj = format.deserialize(req, argType);
+    } catch (Throwable t) {
+      logger.log(Level.SEVERE, "Failed to parse request for " + function.getClass().getName(), t);
+      res.setStatusCode(HttpServletResponse.SC_BAD_REQUEST);
+      return;
+    }
+
+    Object resObj = null;
+    try {
+      resObj = function.apply(reqObj);
+    } catch (Throwable t) {
+      logger.log(Level.SEVERE, "Failed to execute " + function.getClass().getName(), t);
+      res.setStatusCode(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+      return;
+    }
+
+    try {
+      format.serialize(resObj, res);
+    } catch (Throwable t) {
+      logger.log(
+          Level.SEVERE, "Failed to serialize response for " + function.getClass().getName(), t);
+      res.setStatusCode(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+      return;
+    }
+  }
+
+  private static class ConfigurationImpl implements TypedFunction.Configuration {
+    private WireFormat format = null;
+
+    WireFormat getFormat() {
+      if (format == null) {
+        synchronized (this) {
+          if (format == null) {
+            format = new GsonWireFormat();
+          }
+        }
+      }
+
+      return format;
+    }
+
+    @Override
+    public Configuration setWireFormat(WireFormat format) {
+      this.format = format;
+      return null;
+    }
+  }
+
+  private static class GsonWireFormat implements TypedFunction.WireFormat {
+    private final Gson gson = new GsonBuilder().create();
+
+    @Override
+    public void serialize(Object object, HttpResponse response) throws Exception {
+      if (object == null) {
+        response.setStatusCode(204);
```

Comment:
A comment would probably be helpful here, or you could use `HttpServletResponse.SC_NO_CONTENT`.